crates/bevy_ecs/src/schedule/set.rs

```diff
@@ -112,6 +114,58 @@ impl<T> SystemSet for SystemTypeSet<T> {
     }
 }
 
+/// A [`SystemSet`] implicitly created when using
+/// [`Schedule::add_systems`](super::Schedule::add_systems).
+///
+/// The names of the members are stored in a reference counted slice.
+/// The pointer to the slice is used as hash value and for the equality checks.
+pub struct AnonymousSystemSet {
+    system_names: Arc<[Cow<'static, str>]>,
+}
+
+impl AnonymousSystemSet {
+    pub(crate) fn new(system_names: impl Iterator<Item = Cow<'static, str>>) -> Self {
+        Self {
+            system_names: system_names.collect(),
+        }
+    }
+}
+
+impl Debug for AnonymousSystemSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "({})", self.system_names.join(", "))
+    }
+}
+
+impl Hash for AnonymousSystemSet {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Arc::as_ptr(&self.system_names).hash(state);
```

Comment:
This needs a comment explaining the details you captured in the PR description. This is critical logic, and easy to overlook.

Reply:
Okay, I have added comments to both the Hash and PartialEq implementation and moved the PartialEq implementation closer to the Hash implementation to make it harder to overlook.